rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,894 @@
+                    clamp_to_usize(chunk_state.compressed_remaining),
+                );
+                // if we have cleared the compressed data buffer, reset it to 0 instead of infinitely growing
+                if self.compressed_data_start == self.compressed_data_end {
+                    let empty_bytes = self.compressed_data.len() - self.compressed_data_end;
+                    self.compressed_data.resize(empty_bytes, 0);
+                    self.compressed_data_start = 0;
+                    self.compressed_data_end = 0;
+                }
+            }
+        }
+        if slice_end <= self.uncompressed_data_end {
+            return Ok(Bounds((self.uncompressed_data_start, slice_end)));
+        }
+        Ok(Remainder(slice_end - self.uncompressed_data_end))
+    }
+
+    // Consume `amount` bytes of the uncompressed input buffer if enough is available. On failure,
+    // return the extra amount required as an error value.
+    fn consume(&mut self, amount: usize) -> McapResult<BoundsOrRemainder> {
+        match self.load(amount)? {
+            Bounds(bounds) => {
+                self.uncompressed_data_start += amount;
+                Ok(Bounds(bounds))
+            }
+            Remainder(remainder) => Ok(Remainder(remainder)),
+        }
+    }
+
+    // Return an InputBuf that requests `want` uncompressed bytes from the input file. If reading
+    // from a chunk, requests the amount hinted by the decompressor on the previous iteration.
+    fn request(&mut self, want: usize) -> McapResult<Option<ReadAction>> {
+        let desired_end = self.uncompressed_data_end + want;
+        self.uncompressed_data
+            .resize(std::cmp::max(self.uncompressed_data.len(), desired_end), 0);
+
+        return match &self.from {
+            File => Ok(Some(ReadAction::Fill(InputBuf {
+                buf: &mut self.uncompressed_data[self.uncompressed_data_end..desired_end],
+                total_filled: &mut self.uncompressed_data_end,
+                at_eof: &mut self.at_eof,
+                data_section_hasher: &mut self.data_section_hasher,
+            }))),
+            Chunk(chunk_state) => {
+                let desired_compressed_end = self.compressed_data_end + chunk_state.next_read_size;
+                self.compressed_data.resize(
+                    std::cmp::max(self.compressed_data.len(), desired_compressed_end),
+                    0,
+                );
+                Ok(Some(ReadAction::Fill(InputBuf {
+                    buf: &mut self.compressed_data
+                        [self.compressed_data_end..desired_compressed_end],
+                    total_filled: &mut self.compressed_data_end,
+                    at_eof: &mut self.at_eof,
+                    data_section_hasher: &mut self.data_section_hasher,
+                })))
+            }
+        };
+    }
+}
+
+impl Default for LinearReader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub enum ReadAction<'a> {
```

Comment:
I feel like this could have a better name.

Reply:
ReadOperation? McapReadStep? ReaderState?

I guess ReadAction is fine.